libsolidity/Types.cpp

```diff
@@ -853,6 +853,15 @@ TypePointer ArrayType::externalType() const
 		return std::make_shared<ArrayType>(DataLocation::Memory, baseExt, m_length);
 }
 
+u256 ArrayType::memorySize() const
+{
+	solAssert(!isDynamicallySized(), "");
+	solAssert(m_location == DataLocation::Memory, "");
+	u256 size =  m_length * m_baseType->memoryHeadSize();
```

Comment:
If `m_length * m_baseType->memoryHeadSize()` does not fit into `u256`, it is wrapped. This is quite unlikely, but still, it might be possible to bypass some gas calculations.
Perhaps use
```
bigint size = bigint(m_length) * m_baseType->memoryHeadSize();
solAssert(size <= numeric_limits<u256>::max(), "Array size does not fit u256.");
return u256(size);
```
This might still overflow for the subsequent add where memorySize is used, though.